src/formatter.rs

```diff
@@ -6,48 +6,128 @@ use crate::outputstyle::OutputStyle;
 
 // use lazy_static! to make sure regexes are only compiled once
 lazy_static! {
+    // NOTE
+    // [[VARIABLE]] - {{CONSTANT}} - <<BOLD>> - `MONOSPACE`
     static ref RE_VARIABLE: Regex = Regex::new(r"\[\[((?s).*?)\]\]").unwrap();
     static ref RE_CONSTANT: Regex = Regex::new(r"\{\{((?s).*?)\}\}").unwrap();
     static ref RE_BOLD: Regex = Regex::new(r"<<((?s).*?)>>").unwrap();
     static ref RE_MONOSPACE: Regex = Regex::new(r"`([^`]*?)`").unwrap();
-    static ref RE_MONOSPACE_TRIM: Regex = Regex::new(r"\n? *(`[^`]*`) *").unwrap();
+    static ref RE_MONOSPACE_OLD: Regex = Regex::new(r"```([^`]*?)```").unwrap();
+    static ref RE_MONOSPACE_TRIM: Regex = Regex::new(r"\s*`(?: *\n)?([^`]+?)\s*`\s*").unwrap();
     static ref RE_BACKTICK: Regex = Regex::new(r"(`[^`]+`)|([^`]+)").unwrap();
     static ref RE_SPACES: Regex = Regex::new(r" +").unwrap();
     static ref RE_NONWHITESPACE: Regex = Regex::new(r"[^\r\n ]+").unwrap();
+    static ref RE_NEWLINES: Regex = Regex::new(r"\n\n\n+").unwrap();
+    static ref RE_ALL_BUT_MONOSPACE: Regex =
+        Regex::new(r"\[\[((?s).*?)\]\]|\{\{((?s).*?)\}\}|<<((?s).*?)>>").unwrap();
 }
 
-/// Format Codingame statement that contains special formatting syntax
-/// [[VARIABLE]] - {{CONSTANT}} - <<BOLD>> - `MONOSPACE`
 pub fn format_cg(text: &str, ostyle: &OutputStyle) -> String {
-    // Trim consecutive spaces (imitates html behaviour)
-    // But only if it's not in a Monospace block (between backticks ``)
-    let mut result = RE_BACKTICK
-        .replace_all(text, |caps: &regex::Captures| {
-            if let Some(backtick_text) = caps.get(1) {
-                backtick_text.as_str().to_string()
-            } else if let Some(non_backtick_text) = caps.get(2) {
-                RE_SPACES.replace_all(non_backtick_text.as_str(), " ").to_string()
+    if RE_MONOSPACE_OLD.is_match(&text) {
+        println!(
+            "{} {}\n",
+            ostyle.failure.paint("WARNING"),
+            "Clash contains obsolete ``` formatting, consider fixing it in the website."
+        );
+    }
+
+    let mut text = format_edit_monospace(&text);
+    text = format_trim_consecutive_spaces(&text);
+    text = format_monospace_padding(&text);
+    text = format_add_reverse_nester_tags(&text);
+    text = format_paint_inner_blocks(&text, ostyle);
+    format_remove_excessive_newlines(&text)
+}
+
+/// 1. Replaces ```text``` -> `text`
```

Comment:
This and various other doc comments still don't look right because they contain markdown (note how both \`\`\`text\`\`\` and \`text\` look the same in the rendered hover text):

![image](https://github.com/Andriamanitra/clash/assets/10672443/83733f86-062b-4f26-8696-e004003612ed)